aframe/tasks/data/waveforms.py

```diff
@@ -45,40 +78,117 @@ def run(self):
         if not detector_frame_prior:
             samples = convert_to_detector_frame(samples)
         signals = generator(samples)
         with self.output().open("w") as f:
-            print(f, f.path, f.tmp_path)
             write_waveforms(f, signals, samples, generator)
 
 
-# for validation waveforms, utilize rejection sampling
-# to generate waveforms with same distribution as testing set
 @inherits(WaveformParams)
-class ValidationWaveforms(law.Task):
-    ifos = luigi.ListParameter()
-    snr_threshold = luigi.FloatParameter()
-    highpass = luigi.FloatParameter()
+class DeployValidationWaveforms(
+    AframeDataTask,
+    law.LocalWorkflow,
+    StaticMemoryWorkflow,
+):
+    """
+    Generate waveforms for validation via rejection sampling
+    """
+
+    output_dir = luigi.Parameter(
+        description="Directory where validation waveforms will be saved"
+    )
+    ifos = luigi.ListParameter(
+        description="Interferometers for which waveforms will be generated"
+    )
+    snr_threshold = luigi.FloatParameter(
+        description="SNR threshold for rejection sampling"
+    )
+    highpass = luigi.FloatParameter(
+        description="Frequency of highpass filter in Hz"
+    )
+    num_jobs = luigi.IntParameter(
+        default=10,
+        description="Number of parallel jobs "
+        "to split waveform generation amongst",
+    )
+
+    def requires(self):
+        reqs = {}
+        reqs["psd_segment"] = FetchTrain.req(
+            self,
+            branch=-1,
+            data_dir=os.path.join(self.output_dir, "background"),
+            segments_file=os.path.join(self.output_dir, "segments.txt"),
+        )
+        return reqs
+
+    def workflow_requires(self):
+        reqs = super().workflow_requires()
+        reqs["psd_segment"] = FetchTrain.req(
+            self,
+            branch=-1,
+            data_dir=os.path.join(self.output_dir, "background"),
+            segments_file=os.path.join(self.output_dir, "segments.txt"),
+        )
+        return reqs
 
     @property
-    def psds(self):
-        raise NotImplementedError
+    def tmp_dir(self):
+        return os.path.join(self.output_dir, f"tmp-{self.branch}")
 
     def output(self):
-        return law.LocalFileTarget(self.output_file)
+        return law.LocalFileTarget(
+            os.path.join(self.tmp_dir, "val_waveforms.hdf5")
+        )
+
+    def create_branch_map(self):
+        # split the number of signals into num_jobs branches
+        # TODO: account for remainder
+        waveforms_per_branch, remainder = divmod(
+            self.num_signals, self.num_jobs
+        )
+        return {i: waveforms_per_branch for i in range(self.num_jobs)}
+
+    @property
+    def psd_segment(self):
+        """
+        Segment to use to calculate psd for rejection sampling
+        """
+        return list(self.input()["psd_segment"].collection.targets.values())[
+            -1
+        ]
 
     def run(self):
+        import io
```

Comment:
Typically yes - briefly, `law` adds the ability for luigi `Tasks` to be run inside specified environments. This could be apptainer images, or even venv/poetry/python environments. This is analogous to how `pinto` would activate specific `conda`/`poetry` environments before running a task

What this means is that the necessary imports to run a Task are (likely) not available in the environment you use to launch and build the task dependency graph, so you import them inside `run` which will be called once the corresponding Task environment has been activated.